File: app/src/main/java/com/vladislav/workoutassistant/utilities/DiffUtilCallback.java
package com.vladislav.workoutassistant.utilities;

import com.vladislav.workoutassistant.data.models.Identifiable;

import java.util.List;

import androidx.recyclerview.widget.DiffUtil;

public class DiffUtilCallback extends DiffUtil.Callback {

    private List<? extends Identifiable> mOldList;
    private List<? extends Identifiable> mNewList;

    public DiffUtilCallback(List<? extends Identifiable> oldList, List<? extends Identifiable> newList) {
        mOldList = oldList;
        mNewList = newList;
    }

    @Override
    public int getOldListSize() {
        return mOldList.size();
    }

    @Override
    public int getNewListSize() {
        return mNewList.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return mOldList.get(oldItemPosition).getId() == mNewList.get(newItemPosition).getId();
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        return mOldList.get(oldItemPosition).equals(mNewList.get(newItemPosition));
    }
}

File: app/src/main/java/com/vladislav/workoutassistant/utilities/DiffUtilCallback.kt
package com.vladislav.workoutassistant.utilities

import com.vladislav.workoutassistant.data.models.Identifiable

import androidx.recyclerview.widget.DiffUtil

class DiffUtilCallback(private val mOldList: List<Identifiable>, private val mNewList: List<Identifiable>) : DiffUtil.Callback() {

    override fun getOldListSize(): Int = mOldList.size

    override fun getNewListSize(): Int = mNewList.size

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean =
            mOldList[oldItemPosition].id == mNewList[newItemPosition].id

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean =
            mOldList[oldItemPosition] == mNewList[newItemPosition]
}
